src/Runtime/OMSort.inc

```diff
@@ -0,0 +1,460 @@
+// before entering the loop.
+//
+// This function is expected to provide "stable" sort that preserve the input
+// data order, if the values are the same. Although most sort algorithms (e.g.
+// quick sort) are not stable, we use special compare functions to compare two
+// values at first, then compare the input orders if the values are the same.
+// This comparison functions guarantee the input order among the same values
+// and makes sorting algorithms stable.
+//
+#if defined(__APPLE__)
+typedef int(
+    compareFunctionType(void *dataPtr, const void *idx1, const void *idx2));
+#define declare_compare_function(fname, typeName, direction, symbol)           \
+  int compare##fname##direction(                                               \
+      void *dataPtr, const void *idx1p, const void *idx2p) {                   \
+    uint64_t idx1 = *((uint64_t *)idx1p);                                      \
+    uint64_t idx2 = *((uint64_t *)idx2p);                                      \
+    typeName *data = (typeName *)dataPtr;                                      \
+    return ((data[idx1] symbol data[idx2]) ||                                  \
+               ((data[idx1] == data[idx2]) && (idx1 > idx2)))                  \
+               ? 1                                                             \
+               : -1;                                                           \
+  }
+#elif defined(_MSC_VER)
+typedef int(__cdecl compareFunctionType)(void *, const void *, const void *);
+#define declare_compare_function(fname, typeName, direction, symbol)           \
+  int compare##fname##direction(                                               \
+      void *dataPtr, const void *idx1p, const void *idx2p) {                   \
+    uint64_t idx1 = *((uint64_t *)idx1p);                                      \
+    uint64_t idx2 = *((uint64_t *)idx2p);                                      \
+    typeName *data = (typeName *)dataPtr;                                      \
+    return (data[idx1] symbol data[idx2]) ||                                   \
+           ((data[idx1] == data[idx2]) && (idx1 > idx2));                      \
+  }
+#else
+typedef int(
+    compareFunctionType(const void *idx1, const void *idx2, void *dataPtr));
+#define declare_compare_function(fname, typeName, direction, symbol)           \
+  int compare##fname##direction(                                               \
+      const void *idx1p, const void *idx2p, void *dataPtr) {                   \
+    uint64_t idx1 = *((uint64_t *)idx1p);                                      \
+    uint64_t idx2 = *((uint64_t *)idx2p);                                      \
+    typeName *data = (typeName *)dataPtr;                                      \
+    return (data[idx1] symbol data[idx2]) ||                                   \
+           ((data[idx1] == data[idx2]) && (idx1 > idx2));                      \
+  }
+#endif
```

Comment:
Windows and Linux cases are missing the fix for making qsort stable.

Reply:
@gongsu832 The "? 1 : -1" part is comes from comparison function specification only for MacOS, the compare functions for Windows and Linux should return 0 if comparison fails.

Reply:
Not sure where you get that. On all platforms, the comparison function should return < 0, = 0, or > 0 if the 1st element is less than, equal to, or larger than the 2nd element. Of course, to make it stable, you can return > 0 or < 0 when the two elements are equal.

Reply:
@gongsu832 I am sorry. I fixed the issues. Thank you!